### DISCUSSION TITLE
useDrag - hook to drag elements around

### DISCUSSION THREAD

**Discussion by ruiramos** [mantinedev/mantine#3869] [Feature requests]
Hi, 
Using mantine for a short while but really enjoying it so far, awesome work everyone. 🙏
I have a requirement of making an element movable on the page, being able to drag it somewhere else with the mouse, and I've noticed there's no hook to do this easily. Would this be something you'd be looking to add? Happy to contribute.
Thanks,
Rui

**Comment by rtivital** [comment]
It is a complex hook, I'd rather not maintain it as a part of the library for now. You can user a third-party library for it.

**Reply by ruiramos** [reply]
Hey @rtivital thanks for getting back to me. You might think it's a naive implementation at points, but it took me not that many lines of code to get it working:

```
const useDrag = () => {
  const ref = useRef<HTMLDivElement>(null);
  const [dragging, setDragging] = useState(false);
  const initialPos = useRef({ x: 0, y: 0, objX: 0, objY: 0 });

  const onMouseDown = (e: MouseEvent) => {
    if (!ref.current) return;
    setDragging(true);
    initialPos.current = {
      x: e.clientX,
      y: e.clientY,
      objX: parseInt(ref.current.style.left),
      objY: parseInt(ref.current.style.top),
    };
  };

  const onMouseMove = (e: MouseEvent) => {
    if (!ref.current) return;
    const initial = initialPos.current;

    const xDelta = e.clientX - initial.x;
    const yDelta = e.clientY - initial.y;

    ref.current.style.left = `${initial.objX + xDelta}px`;
    ref.current.style.top = `${initial.objY + yDelta}px`;
  };

  const onMouseUp = () => {
    setDragging(false);
  };

  // disable userSelect so text doesn't get selected when dragging
  useEffect(() => {
    if (dragging) {
      window.document.body.style.userSelect = "none";
    } else {
      window.document.body.style.userSelect = "";
    }
  }, [dragging]);

  useEffect(() => {
    if (!ref.current) return;
    const currentRef = ref.current;
    ref.current.addEventListener("mousedown", onMouseDown);

    if (dragging) {
      window.addEventListener("mousemove", onMouseMove);
      window.addEventListener("mouseup", onMouseUp);
    } else {
      window.removeEventListener("mousemove", onMouseMove);
      window.removeEventListener("mouseup", onMouseUp);
    }

    return () => {
      if (!currentRef) return;
      currentRef.removeEventListener("mousedown", onMouseDown);
      window.removeEventListener("mousemove", onMouseMove);
      window.removeEventListener("mouseup", onMouseUp);
    };
  }, [ref, dragging]);

  return { ref, active: dragging };
};
```

Up to you!

**Reply by rtivital** [reply]
There are usually some caveats, for example `use-local-storage` hook does not seems that hard – you just need to save given value to the local storage. But then, there appear a lot of additional issues that should be added:
- values syncing between hooks that use the same key
- values syncing between browser tabs
- session storage support
- value serialization/deserialization

It all stacks up over time and hook starts looking like this – https://github.com/mantinedev/mantine/blob/master/src/mantine-hooks/src/use-local-storage/create-storage.ts

**Reply by ruiramos** [reply]
Fair enough! This stuff is hard! :) 
Feel free to close this thread.
